### ISSUE TITLE
Auto/manual save within editing mode or narrow mode

### ISSUE THREAD

**Issue by dalbandea** [amake/orgro#126]
Currently autosave only seems to occur when exiting edit mode and all narrow modes, which I find a bit inconvenient: if I want to save and keep editing, I must exit edit mode, narrow mode, and then look again for the section I was editing to get into narrow mode and edit mode again. Also, many times I want to switch apps while editing within a narrow mode and there's the risk that the app gets closed and changes are lost.

I think it would be better if autosave is triggered every time one exits from edit mode. Maybe it would also be convenient to have a manual save option within edit mode?

**Comment by amake** [comment]
> Currently autosave only seems to occur when exiting edit mode and all narrow modes

This is a consequence of the current architecture. Changes in the narrowed view aren't "applied" to the original file until you exit the narrowed view. Narrowed views are not considered "saveable" because they are completely divorced from the context of their originating file.

> Also, many times I want to switch apps while editing within a narrow mode and there's the risk that the app gets closed and changes are lost

This shouldn't happen. I have taken precautions to ensure that the narrow state and editing state are preserved and restored even if the app is ejected from memory. If that's not working for you then please let me know so I can fix it. (I did just discover that after restoration you must make at least one additional edit to the content in order for it to save correctly; this will be fixed in the next version.)

Do you really need to be able to save while narrowed, or are you just concerned about data loss?

**Comment by amake** [comment]
The fix for applying an edit immediately after restoration will be available in v1.50.2, ready for testing soon:

- [iOS](https://testflight.apple.com/join/8YTxcuZC)
- [Android](https://play.google.com/store/apps/details?id=com.madlonkay.orgro)

**Comment by dalbandea** [comment]
I'm concerned only about data loss, the manual save functionality isn't necessary for me if data integrity is ensured. If I'm in edit mode and do some edit, and I close the app from the recent apps (removing it from memory), when I open it again I see the "Recent files" view

**Comment by amake** [comment]
Manually removing it from memory is different from being ejected from memory by the system. State restoration only works on the latter. To test it you can enable the "Don't keep activities" system debug setting: https://developer.android.com/studio/debug/dev-options

0. Enable "Don't keep activities"
1. Open the editor in Orgro and change the content
2. Switch away from Orgro
3. Return to Orgro. The editor state should be restored. From 1.50.2 saving immediately will work; in earlier versions you need to make an additional change to the text.

**Comment by dalbandea** [comment]
Yes, that should be working. The problem I see is that it's inconvenient to need to get out of editing and narrowed mode so that the save is applied, as I'm used to compulsively saving progress in emacs and vim on the laptop. Also, it can happen that when doing app switching while narrowed after editing I forget to come back to save progress and remove the app from memory manually.

**Comment by amake** [comment]
There are various architectural problems that make it difficult to allow persisting to disk when narrowed. My feeling is that thorough and rigorous state restoration is sufficient, so I am not inclined to look into this further at the moment.

> I forget to come back to save progress and remove the app from memory manually

Manually removing apps from memory like that is not recommended (see [here](https://www.androidcentral.com/apps-software/should-you-force-close-your-apps), [here](https://www.slashgear.com/910451/why-you-need-to-stop-closing-apps-on-your-android-phone/), [here](https://www.howtogeek.com/771413/stop-closing-apps-on-your-android-phone/), etc.) so in my opinion this is borderline operator error.

**Comment by SteepturN** [comment]
@amake maybe you can add saving narrowed text to a _view file_ in the app directory for extra safety? After thorough and rigorous state restoration and saving the initial file, app could delete _view file_. If something happens, app or user could restore changed content.

There are situations in which exiting isn't an operator error like OS crash, battery discharge.

**Comment by amake** [comment]
> There are situations in which exiting isn't an operator error like OS crash, battery discharge.

And in those cases the existing restoration framework should work fine. Is that not the case? Can you point to a specific scenario where restoration does not work as expected? (Please include steps to reproduce.)

> maybe you can add saving narrowed text to a view file in the app directory for extra safety?

I don't see the point of adding another store for restoration data, unless there is some concrete scenario that *can't* be covered by the existing framework.

**Comment by farynaio** [comment]
I agree with OP, I also found it confusing, and inconvenient to need to close narrow mode to make sure edition is saved. I sync my notes between various devices, and edit on all of them, so making sure the content is persisted on the disk asap is mandatory to avoid conflicts.

Ideally there would be icon to save edition on demand. In narrow mode, it could work in similar way as `save-excursion`: temporarily persist current narrow state > leave all narrow states > save edition > restore narrow state.
